cookbooks/bcpc-hadoop/libraries/helpers.rb

```diff
@@ -0,0 +1,66 @@
+
+module Bcpc_Hadoop
+  module Helper
+
+    include Chef::Mixin::ShellOut
+
+    # Apply translation rules for Hortonworks release strings
+    # to the debian package name strings
+    #
+    # release - String dotted Hortonworks release (e.g. 2.2.0-2041)
+    # package name - String for base package name
+    #
+    # Raises RuntimeError on any unspecified error
+    # Returns - string to use in a Hortonworks debian package name
+    # (i.e. 2.2.0.2041 for a release and package like
+    #  hadoop-hdfs-datanode to something akin to
+    #  hadoop-2-2-0-2041-hdfs-datanode)
```

Comment:
This is Hortonworks specific as I understand it. Cloudera or Apache Bigtop may adopt it but I am unsure how wide spread it is. This allows for rolling-upgrade on a machine. See the Hadoop Summit 2015 [presentation on rolling upgrade](https://www.youtube.com/watch?v=sSL7v1HcmWw) for a bit more if I recall correctly.